src/PrefixTree.java

```diff
@@ -30,55 +27,51 @@ public boolean contains(CharSequence word){
             System.out.println("Word has been added already. " + word);
             return false;
         }
-        else{
-            for (int i = helper.index; i < word.length(); i++) {
-                if(word.length() == 1){
-                    break;
-                }
-                Edge edge = new Edge();
-                Node node = new Node();
-                node.setValue(word.charAt(i));
-                edge.setReference(node);
-                previousNode.addEdge(edge);
-                previousNode = node;
-            }
-            previousNode.setFinal(true);
+        if(word.length() == 1){
+            currentNode.setFinal(true);
             return true;
         }
+        for (int i = helper.index; i < word.length(); i++) {
+            Node nextNode = new Node(word.charAt(i));
+            Edge edge = new Edge(nextNode);
+            currentNode.addEdge(edge);
+            currentNode = nextNode;
+            }
+            currentNode.setFinal(true);
+            return true;
     }
 
-    private Helper traverse(CharSequence word){
-        Helper helper = new Helper();
-        Node previousNode = root;
-        for (int i = 0; i < word.length(); i++) {
-            boolean hasEdge = false;
-            ArrayList<Edge> nodeEdges = previousNode.getEdges();
-            for (Edge nodeEdge : nodeEdges) {
-                if(nodeEdge.getReferences().getValue() == word.charAt(i)){
-                    previousNode = nodeEdge.getReferences();
-                    hasEdge = true;
-                    break;
-                }
-            }
-            if(!hasEdge){
-                helper.index = i;
-                helper.node = previousNode;
-                return helper;
+    private TraverseResult traverse(CharSequence word){
+        TraverseResult traverseResult = new TraverseResult();
+        traverseResult.node = root;
+        for (traverseResult.index = 0; traverseResult.index < word.length(); traverseResult.index++) {
+            AtomicBoolean hasEdge = new AtomicBoolean(false);
+            ArrayList<Edge> nodeEdges = traverseResult.node.getEdges();
+
+            nodeEdges.forEach(
+                    edge -> {
+                        if (edge.getNode().getValue() == word.charAt(traverseResult.index)) {
+                            traverseResult.node = edge.getNode();
+                            hasEdge.set(true);
+                        }
+                    }
+            );
+
+            if(!hasEdge.get()){
+                return traverseResult;
             }
         }
-        if(!previousNode.isFinal()){
-            helper.index = word.length() - 1;
-            helper.node = previousNode;
-            return helper;
+        if(!traverseResult.node.isFinal()){
+            traverseResult.index = word.length() - 1;
+            return traverseResult;
         }
-        helper.index = -1;
-        helper.node = previousNode;
-        return helper;
+        traverseResult.index = -1;
```

Comment:
All of this can go away really.  We can change the contract here and instead of `index` have the number of characters that got matched. The root node is never final, so `contains` will just need to check if the returned node is final (as well as the processed number of characters), while `add` should be adjusted slightly.